feat: reuse existing model downloading tasks

axel7083 opened the pull request:
### What does this PR do?

When we start a model download in the models Table, we have an issue when starting the application, where it detects the models already being on disk (in reality it is **partially** on disk) and it skip the step, when it should not.

Now, when requesting to download a model, the system will check for existing downloader and tasks. Avoiding potential errors.

### Screenshot / video of UI

https://github.com/projectatomic/ai-studio/assets/42176370/4d068def-8323-40bb-a905-dc52a5d03e3a

### What issues does this PR fix or reference?

Fixes https://github.com/projectatomic/ai-studio/issues/384

### How to test this PR?

- [x] Unit tests has been provided

(1) Go to models table 
(2) Download a model
(3) Go to any recipe that can use this model
(4) Click on "Start application"

lstocchi commented on packages/backend/src/managers/modelsManager.ts:
```diff
@@ -171,7 +173,101 @@ export class ModelsManager implements Disposable {
+  }
+
+  private createDownloader(model: ModelInfo): Downloader {
+    console.log('Creating a downloader.');
```
to be removed

lstocchi commented on packages/backend/src/managers/modelsManager.ts:
```diff
@@ -171,7 +173,101 @@ export class ModelsManager implements Disposable {
+  async requestDownloadModel(model: ModelInfo, labels?: { [key: string]: string }): Promise<string> {
+    // Check there is no existing downloader running
+    if (!this.#downloaders.has(model.id)) {
+      console.debug('no downloader has been found.');
```
remove

lstocchi reviewed (commented):
Code LGTM, a couple of console log to be removed. Gonna give it a try

axel7083 commented:
> If i try to start 2 recipes (summarizer and code gen) with the same model, the first works fine i see the model downloading progressing (and also in the models page), on the second recipe i just see the checkout repository status and stop. Like it is stuck
> 
> ![image](https://private-user-images.githubusercontent.com/49404737/306715224-5ec9f939-ee35-42f4-81f7-1d637b7d70fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1OTM1NDEsIm5iZiI6MTcwODU5MzI0MSwicGF0aCI6Ii80OTQwNDczNy8zMDY3MTUyMjQtNWVjOWY5MzktZWUzNS00MmY0LTgxZjctMWQ2MzdiN2Q3MGZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIyVDA5MTQwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyZTQ2NTUxMzFjZmI5MjI4NzE3ZGM0MjYyMzhmODNiYmJlMTYwMjY1NmNjNzYyNjY4MzkzZTFhNDM1Y2VhMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ObIttOBuhlQAJ-qRB1FDE1F6ke7MwKb_srifjX_34jc)
> 
> when the model is downloaded (i see it in the models page) an error popup ![image](https://private-user-images.githubusercontent.com/49404737/306715464-abba0d5e-395a-47b9-be04-bae3f68c93ce.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1OTM1NDEsIm5iZiI6MTcwODU5MzI0MSwicGF0aCI6Ii80OTQwNDczNy8zMDY3MTU0NjQtYWJiYTBkNWUtMzk1YS00N2I5LWJlMDQtYmFlM2Y2OGM5M2NlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIyVDA5MTQwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlYTY2OGUwMTk5YjI2NzEwMmRmYzE3YmEyNGRmMzY4NTY0MzI5ZjZhYTY3ODliMjZjZTkyMzJiODMxYTQ0MWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PiBlCUi6IXsCtXe_O9XyAeM2WmpwcUKeJ9NwkLY1q-I) if i go to the containers page i see the pod has been created but if i start it it says libpod error

That's the **very very** annoying case. I think we should not allow starting simultaneously multiple recipe at _the same time_.

jeffmaury commented:
I don't understand why we can start multiple recipe using the same model. When a task is found for the model, instead of updating it for the recipe, can't we add a new task for recipe_id, model_id ?

axel7083 commented:
> I don't understand why we can start multiple recipe using the same model. When a task is found for the model, instead of updating it for the recipe, can't we add a new task for recipe_id, model_id ?

We can, i will work on that this afternoon

axel7083 commented:
@lstocchi @jeffmaury thanks for your feedback, I updated the implementation to properly manage multiple AI-App being downloaded at the same time. I updated the video demonstration in the pull request.
